cas-server-support-pac4j/src/main/java/org/jasig/cas/support/pac4j/authentication/handler/support/AbstractWrapperAuthenticationHandler.java

```diff
@@ -0,0 +1,116 @@
+/*
+ * Licensed to Apereo under one or more contributor license
+ * agreements. See the NOTICE file distributed with this work
+ * for additional information regarding copyright ownership.
+ * Apereo licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file
+ * except in compliance with the License.  You may obtain a
+ * copy of the License at the following location:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.jasig.cas.support.pac4j.authentication.handler.support;
+
+import org.jasig.cas.authentication.Credential;
+import org.jasig.cas.authentication.HandlerResult;
+import org.jasig.cas.authentication.PreventedException;
+import org.jasig.cas.support.pac4j.authentication.principal.ClientCredential;
+import org.pac4j.core.credentials.Credentials;
+import org.pac4j.core.exception.CredentialsException;
+import org.pac4j.core.profile.UserProfile;
+import org.pac4j.core.util.CommonHelper;
+import org.pac4j.http.credentials.authenticator.Authenticator;
+import org.pac4j.http.profile.creator.AuthenticatorProfileCreator;
+import org.pac4j.http.profile.creator.ProfileCreator;
+
+import javax.security.auth.login.FailedLoginException;
+import java.security.GeneralSecurityException;
+
+/**
+ * Abstract pac4j authentication handler which uses a pac4j authenticator and profile creator.
+ *
+ * @author Jerome Leleu
+ * @since 4.2.0
+ */
+@SuppressWarnings("unchecked")
+public abstract class AbstractWrapperAuthenticationHandler extends AbstractPac4jAuthenticationHandler {
+
+    /**
+     * The pac4j authenticator used for authentication.
+     */
+    protected Authenticator authenticator;
```

Comment:
Is it possible to auto detect what authenticators are available in the context? I think you need a map here for every authenticator to be linked to its profile creator. and you can then put the map into the context. entries in the map can be declared dynamically via Components. 

Makes sense?